[WIP][quantization] Use torchbind for Linear PackedParams

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32938 [WIP][quantization] Use torchbind for Linear PackedParams**

Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

Commit:
[WIP][quantization] Use torchbind for PackedParams

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit fb6876af:


* **6/6** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=fb6876af4f709e761dd66ca297f3bc7d183f1004).

### :detective: 6 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4404369) pytorch_linux_xenial_py3_clang5_mobile_build (1/6)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94738521) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94738521))
<details>
<summary>
<code>Feb 03 19:57:40 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:59:58: error: 'bias' was not declared in this scope</code>
</summary>

```
Feb 03 19:57:40        static_cast<fbgemm::float16*>(unpacked_weight.data_ptr()), 
Feb 03 19:57:40                          ^ 
Feb 03 19:57:40 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:56:25: error: expected '(' before '::' token 
Feb 03 19:57:40 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:56:25: error: '::float16' has not been declared 
Feb 03 19:57:40 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:56:35: error: expected primary-expression before '>' token 
Feb 03 19:57:40        static_cast<fbgemm::float16*>(unpacked_weight.data_ptr()), 
Feb 03 19:57:40                                    ^ 
Feb 03 19:57:40 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:57:7: error: 'fbgemm' has not been declared 
Feb 03 19:57:40        fbgemm::matrix_op_t::Transpose); 
Feb 03 19:57:40        ^ 
Feb 03 19:57:40 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:59:58: error: 'bias' was not declared in this scope 
Feb 03 19:57:40    return std::make_tuple(unpacked_weight.to(at::kFloat), bias); 
Feb 03 19:57:40                                                           ^ 
Feb 03 19:57:40 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:59:58: note: suggested alternatives: 
Feb 03 19:57:40 In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/Dimname.h:3:0, 
Feb 03 19:57:40                  from /var/lib/jenkins/workspace/aten/src/ATen/core/NamedTensor.h:3, 
Feb 03 19:57:40                  from /var/lib/jenkins/workspace/build_mobile/aten/src/ATen/core/TensorBody.h:19, 
Feb 03 19:57:40                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Feb 03 19:57:40                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
Feb 03 19:57:40                  from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:5, 
Feb 03 19:57:40                  from /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:1: 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4404380) caffe2_onnx_py2_gcc5_ubuntu16_04_build (2/6)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94738598) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94738598))
<details>
<summary>
<code>Feb 03 19:57:53 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:59:58: error: 'bias' was not declared in this scope</code>
</summary>

```
Feb 03 19:57:53        static_cast<fbgemm::float16*>(unpacked_weight.data_ptr()), 
Feb 03 19:57:53                          ^ 
Feb 03 19:57:53 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:56:25: error: expected '(' before '::' token 
Feb 03 19:57:53 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:56:25: error: '::float16' has not been declared 
Feb 03 19:57:53 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:56:35: error: expected primary-expression before '>' token 
Feb 03 19:57:53        static_cast<fbgemm::float16*>(unpacked_weight.data_ptr()), 
Feb 03 19:57:53                                    ^ 
Feb 03 19:57:53 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:57:7: error: 'fbgemm' has not been declared 
Feb 03 19:57:53        fbgemm::matrix_op_t::Transpose); 
Feb 03 19:57:53        ^ 
Feb 03 19:57:53 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:59:58: error: 'bias' was not declared in this scope 
Feb 03 19:57:53    return std::make_tuple(unpacked_weight.to(at::kFloat), bias); 
Feb 03 19:57:53                                                           ^ 
Feb 03 19:57:53 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:59:58: note: suggested alternatives: 
Feb 03 19:57:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/Dimname.h:3:0, 
Feb 03 19:57:53                  from /var/lib/jenkins/workspace/aten/src/ATen/core/NamedTensor.h:3, 
Feb 03 19:57:53                  from /var/lib/jenkins/workspace/build/aten/src/ATen/core/TensorBody.h:19, 
Feb 03 19:57:53                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Feb 03 19:57:53                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
Feb 03 19:57:53                  from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:5, 
Feb 03 19:57:53                  from /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:1: 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4404366) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (3/6)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94738614) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94738614))
<details>
<summary>
<code>Feb 03 19:58:20 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:57:7: error: use of undeclared identifier 'fbgemm'</code>
</summary>

```
Feb 03 19:58:20                ~~~~~~~~~~~~~~~~~  ^ 
Feb 03 19:58:20 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:51:35: error: no member named 'numCols' in 'qnnpack::PackBMatrix' 
Feb 03 19:58:20   auto ncols = packed_weight_ptr->numCols(); 
Feb 03 19:58:20                ~~~~~~~~~~~~~~~~~  ^ 
Feb 03 19:58:20 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:55:22: error: no member named 'unpack' in 'qnnpack::PackBMatrix' 
Feb 03 19:58:20   packed_weight_ptr->unpack( 
Feb 03 19:58:20   ~~~~~~~~~~~~~~~~~  ^ 
Feb 03 19:58:20 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:56:19: error: use of undeclared identifier 'fbgemm' 
Feb 03 19:58:20       static_cast<fbgemm::float16*>(unpacked_weight.data_ptr()), 
Feb 03 19:58:20                   ^ 
Feb 03 19:58:20 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp:57:7: error: use of undeclared identifier 'fbgemm' 
Feb 03 19:58:20       fbgemm::matrix_op_t::Transpose); 
Feb 03 19:58:20       ^ 
Feb 03 19:58:20 19 errors generated. 
Feb 03 19:58:20 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp.o] Error 1 
Feb 03 19:58:20 make[2]: *** Waiting for unfinished jobs.... 
Feb 03 19:58:20 caffe2/CMakeFiles/torch_cpu.dir/build.make:4195: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp.o' failed 
Feb 03 19:58:28 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Feb 03 19:58:28 CMakeFiles/Makefile2:5410: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Feb 03 19:58:28 Makefile:138: recipe for target 'all' failed 
Feb 03 19:58:28 make: *** [all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4404389) pytorch_linux_xenial_py3_clang5_asan_build (4/6)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94738647) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94738647))
<details>
<summary>
<code>Feb 03 19:58:31 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:91:42: error: expected ';' at end of declaration list</code>
</summary>

```
Feb 03 19:58:31            ~ ^ ~~~~~~~~~~~~~~~ 
Feb 03 19:58:31 /usr/include/assert.h:89:5: note: expanded from macro 'assert' 
Feb 03 19:58:31   ((expr)                                                               \ 
Feb 03 19:58:31     ^~~~ 
Feb 03 19:58:31 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:7: 
Feb 03 19:58:31 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:6: 
Feb 03 19:58:31 /var/lib/jenkins/workspace/third_party/fbgemm/include/fbgemm/QuantUtils.h:122:29: warning: comparison of integers of different signs: 'std::size_t' (aka 'unsigned long') and 'int' [-Wsign-compare] 
Feb 03 19:58:31   for (std::size_t i = 0; i < len; i++) { 
Feb 03 19:58:31                           ~ ^ ~~~ 
Feb 03 19:58:31 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:7: 
Feb 03 19:58:31 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:91:42: error: expected ';' at end of declaration list 
Feb 03 19:58:31       int64_t output_zero_point) override [[noreturn]] { 
Feb 03 19:58:31                                          ^ 
Feb 03 19:58:31                                          ; 
Feb 03 19:58:31 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:221:16: warning: unused variable 'qtype' [-Wunused-variable] 
Feb 03 19:58:31     const auto qtype = weight.qscheme(); 
Feb 03 19:58:31                ^ 
Feb 03 19:58:31 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:254:63: warning: comparison of integers of different signs: 'int64_t' (aka 'long') and 'const size_t' (aka 'const unsigned long') [-Wsign-compare] 
Feb 03 19:58:31             (bias_fp32.ndimension() == 1 && bias_fp32.size(0) == out_ch), 
Feb 03 19:58:31                                             ~~~~~~~~~~~~~~~~~ ^  ~~~~~~ 
Feb 03 19:58:31 /var/lib/jenkins/workspace/c10/util/Exception.h:257:31: note: expanded from macro 'TORCH_CHECK' 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4404388) pytorch_macos_10_13_py3_build (5/6)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94738851) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94738851))
<details>
<summary>
<code>Feb 03 11:58:43 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o</code>
</summary>

```
Feb 03 11:58:43 [1758/3006] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/TensorShape.cpp.o 
Feb 03 11:58:43 ../aten/src/ATen/native/TensorShape.cpp:24:24: warning: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int64_t' (aka 'long long') [-Wsign-compare] 
Feb 03 11:58:43   for (size_t i = 0; i < shape_tensor.numel(); ++i) { 
Feb 03 11:58:43                      ~ ^ ~~~~~~~~~~~~~~~~~~~~ 
Feb 03 11:58:43 ../aten/src/ATen/native/TensorShape.cpp:85:25: warning: comparison of integers of different signs: 'int64_t' (aka 'long long') and 'size_t' (aka 'unsigned long') [-Wsign-compare] 
Feb 03 11:58:43   for (int64_t i = 0; i < s1.size(); ++i) { 
Feb 03 11:58:43                       ~ ^ ~~~~~~~~~ 
Feb 03 11:58:43 2 warnings generated. 
Feb 03 11:58:43 [1759/3006] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/TensorTransformations.cpp.o 
Feb 03 11:58:43 [1760/3006] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o 
Feb 03 11:58:43 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o  
_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o -c ../aten/src/ATen/native/RNN.cpp 
Feb 03 11:58:43 In file included from ../aten/src/ATen/native/RNN.cpp:6: 
Feb 03 11:58:43 In file included from ../aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:5: 
Feb 03 11:58:43 ../third_party/fbgemm/include/fbgemm/FbgemmFP16.h:178:24: warning: comparison of integers of different signs: 'uint64_t' (aka 'unsigned long long') and 'int' [-Wsign-compare] 
Feb 03 11:58:43         ((block_row_id != nbrow_ - 1) ? (blockRowSize() * blockColSize()) 
Feb 03 11:58:43           ~~~~~~~~~~~~ ^  ~~~~~~~~~~ 
Feb 03 11:58:43 ../third_party/fbgemm/include/fbgemm/FbgemmFP16.h:185:18: warning: comparison of integers of different signs: 'uint64_t' (aka 'unsigned long long') and 'int' [-Wsign-compare] 
Feb 03 11:58:43     assert(index < matSize()); 
Feb 03 11:58:43            ~~~~~ ^ ~~~~~~~~~ 
Feb 03 11:58:43 /Applications/Xcode-9.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert' 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4404379) pytorch_xla_linux_xenial_py3_6_clang7_build (6/6)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94739247) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94739247))
<details>
<summary>
<code>Feb 03 19:59:31 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:91:42: error: expected ';' at end of declaration list</code>
</summary>

```
Feb 03 19:59:31            ~ ^ ~~~~~~~~~~~~~~~ 
Feb 03 19:59:31 /usr/include/assert.h:89:5: note: expanded from macro 'assert' 
Feb 03 19:59:31   ((expr)                                                               \ 
Feb 03 19:59:31     ^~~~ 
Feb 03 19:59:31 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:7: 
Feb 03 19:59:31 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:6: 
Feb 03 19:59:31 /var/lib/jenkins/workspace/third_party/fbgemm/include/fbgemm/QuantUtils.h:122:29: warning: comparison of integers of different signs: 'std::size_t' (aka 'unsigned long') and 'int' [-Wsign-compare] 
Feb 03 19:59:31   for (std::size_t i = 0; i < len; i++) { 
Feb 03 19:59:31                           ~ ^ ~~~ 
Feb 03 19:59:31 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:7: 
Feb 03 19:59:31 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:91:42: error: expected ';' at end of declaration list 
Feb 03 19:59:31       int64_t output_zero_point) override [[noreturn]] { 
Feb 03 19:59:31                                          ^ 
Feb 03 19:59:31                                          ; 
Feb 03 19:59:31 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:221:16: warning: unused variable 'qtype' [-Wunused-variable] 
Feb 03 19:59:31     const auto qtype = weight.qscheme(); 
Feb 03 19:59:31                ^ 
Feb 03 19:59:31 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:254:63: warning: comparison of integers of different signs: 'int64_t' (aka 'long') and 'const size_t' (aka 'const unsigned long') [-Wsign-compare] 
Feb 03 19:59:31             (bias_fp32.ndimension() == 1 && bias_fp32.size(0) == out_ch), 
Feb 03 19:59:31                                             ~~~~~~~~~~~~~~~~~ ^  ~~~~~~ 
Feb 03 19:59:31 /var/lib/jenkins/workspace/c10/util/Exception.h:257:31: note: expanded from macro 'TORCH_CHECK' 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[WIP][quantization] Use torchbind for PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -197,3 +205,76 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
+                bit_width = std::move(std::get<3>(state));
+
+#ifdef USE_FBGEMM
+                if (backend == "FBGEMM") {
```
we shouldn't serialize the backend and instead use qengine do deside which one to call. You probably don't need branching here and instead you can just invoke the linear_prepack implementation directly, it already does it: https://github.com/pytorch/pytorch/blob/f61b45fc89086d773d099112578bdbc3db13da78/aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp#L180

jamesr66a replied:
I don't think a global qengine is the correct design overall, I've already had to come in and debug issues that that's introduced, and it just seems like it was not a well-thought-out design decision

dzhulgakov commented on test/test_quantized_nn_mods.py:
```diff
@@ -273,15 +273,19 @@ def test_linear_api(self, batch_size, in_features, out_features, use_bias, use_d
-            self.assertEqual(model_dict[key], loaded_dict[key])
+            if isinstance(model_dict[key], torch._C.ScriptObject):
+                assert isinstance(loaded_dict[key], torch._C.ScriptObject)
+                w_model, b_model = torch.ops.quantized.linear_unpack(model_dict[key])
```
you can just call `__getstate__` and then compare the results - it'd also work with convs when you add them

dzhulgakov commented on test/test_quantized_nn_mods.py:
```diff
@@ -465,9 +473,9 @@ def test_linear_api(self, batch_size, in_features, out_features, use_bias, use_f
-                torch.save(qlinear, b)
+            # with self.assertRaisesRegex(RuntimeError, r'torch.save\(\) is not currently supported'):
+            #     b = io.BytesIO()
+            #     torch.save(qlinear, b)
```
does regular torch.save (or pickle) work with the new classes?

dzhulgakov reviewed (commented):
So the untracked header seems to come from importing JIT headers in lite interpeter builds. Probably the easiest way to repro it is to just import custom_class.h header somewhere in ATen - you don't need the rest of this diff - it'd be easier to debug / split dependencies

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D19701952](https://our.internmc.facebook.com/intern/diff/D19701952)

[ghstack-poisoned]
